ligo/stablecoin/operator.ligo

```diff
@@ -0,0 +1,177 @@
+    Layout.convert_from_right_comb ((param.1.1 : operator_tokens))
+} with case operator_tokens_ of
+    All_tokens -> unit
+  | Some_tokens (tokens) -> Set.fold
+      ( function
+          ( const unused   : unit
+          ; const token_id : token_id
+          ) : unit is validate_token_type (token_id)
+      , tokens
+      , unit
+      )
+  end
+
+(*
+ * Helper function that returns `False` if `Some_tokens (Set.empty)`
+ * was provided to operator parameter
+ *)
+function are_tokens_specified
+  ( const param : operator_param
+  ) : bool is case param.1.1 of
+    M_left (u) -> True
+  | M_right (tokens) -> Set.size (tokens) =/= 0n
+  end
+
+(*
+ * Checks whether the given operator is present in provided operator map
+ *)
+function is_operator_impl
+  ( const param     : operator_param
+  ; const operators : operators
+  ) : bool is block
+  { validate_operator_token_type (param)
+  ; const operator_key : (owner * operator) = (param.0, param.1.0)
+  } with case Big_map.find_opt (operator_key, operators) of
+      // Always returns `True` if `Some_tokens (Set.empty)` was passed to the parameter
+      None -> not are_tokens_specified (param)
+    | Some (u) -> True
+    end
+
+(*
+ * Returns a callback of `Is_operator` entrypoint call
+ *)
+function is_operator
+  ( const param     : is_operator_params
+  ; const operators : operators
+  ) : operation is block
+{ const response : is_operator_response =
+    (param.0, is_operator_impl (param.0, operators))
+} with Tezos.transaction (response, 0mutez, param.1)
+
+(*
+ * Validates operators for the given tranfser batch
+ * and operator storage
+ *)
+function validate_operators
+  ( const permissions_descriptor : permissions_descriptor_
+  ; const transfer_params        : transfer_params
+  ; const operators              : operators
+  ) : unit is block
+{ const self_transfer_policy_ : self_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.self : self_transfer_policy))
+
+; const can_transfer_from_self : bool =
+    case self_transfer_policy_ of
+      Self_transfer_permitted (u) -> True
+    | Self_transfer_denied    (u) -> False
+    end
+
+; const operator_transfer_policy_ : operator_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.operator : operator_transfer_policy))
+; const can_transfer_operator : bool =
+    case operator_transfer_policy_ of
+      Operator_transfer_permitted (u) -> True
+    | Operator_transfer_denied    (u) -> False
+    end
+
+; const operator : address = Tezos.sender
+
+; function validate_params
+    ( const param : transfer_param
+    ) : unit is block
+  { var is_operator : bool := False
+  ; const owner : owner = param.0
+  ; if can_transfer_from_self and owner = operator
+    then { is_operator := True }
+    else
+        { const operator_param : operator_param = (owner, (operator, (M_right (set [param.1.1.0]) : operator_tokens)))
+        ; is_operator := is_operator_impl (operator_param, operators)
```

Comment:
But it's still quite inefficient, isn't it? You still validate token_id in `is_operator_impl`, so `transfer` validates it twice. Can't you just do `Big_map.find_opt` here? I don't see a big benefit in extracting `is_operator_impl` here :shrug: